spec/requests/checkout/failed_checkout_spec.rb

```diff
@@ -1,93 +0,0 @@
-  let!(:order_cycle) { create(:simple_order_cycle) }
-  let!(:exchange) {
-    create(:exchange, order_cycle:, sender: order_cycle.coordinator, receiver: shop,
-                      incoming: false, pickup_time: "Monday")
-  }
-  let!(:address) { create(:address) }
-  let!(:line_item) { create(:line_item, order:, quantity: 3, price: 5.00) }
-  let!(:payment_method) {
-    create(:stripe_sca_payment_method, distributor_ids: [shop.id], environment: Rails.env)
-  }
-  let!(:check_payment_method) {
-    create(:payment_method, distributor_ids: [shop.id], environment: Rails.env)
-  }
-  let!(:shipping_method) { create(:shipping_method, distributor_ids: [shop.id]) }
-  let!(:shipment) { create(:shipment_with, :shipping_method, shipping_method:) }
-  let!(:order) {
-    create(:order, shipments: [shipment], distributor: shop, order_cycle:)
-  }
-  let(:params) do
-    { order: {
-      shipping_method_id: shipping_method.id,
-      payments_attributes: [{ payment_method_id: payment_method.id }],
-      bill_address_attributes: address.attributes.slice("firstname", "lastname", "address1",
-                                                        "address2", "phone", "city", "zipcode",
-                                                        "state_id", "country_id"),
-      ship_address_attributes: address.attributes.slice("firstname", "lastname", "address1",
-                                                        "address2", "phone", "city", "zipcode",
-                                                        "state_id", "country_id")
-    } }
-  end
-
-  before do
-    order_cycle_distributed_variants = double(:order_cycle_distributed_variants)
-    allow(OrderCycles::DistributedVariantsService).to receive(:new)
-      .and_return(order_cycle_distributed_variants)
-    allow(order_cycle_distributed_variants)
-      .to receive(:distributes_order_variants?).and_return(true)
-
-    order.reload.update_totals
-    set_order order
-  end
-
-  pending "when shipping and payment fees apply" do
-    let(:calculator) { Calculator::FlatPercentItemTotal.new(preferred_flat_percent: 10) }
-
-    before do
-      payment_method.calculator = calculator.dup
-      payment_method.save!
-      check_payment_method.calculator = calculator.dup
-      check_payment_method.save!
-      shipping_method.calculator = calculator.dup
-      shipping_method.save!
-    end
-
-    it "clears shipments and payments before rendering the checkout" do
-      put update_checkout_path, params:, as: :json
-
-      # Checking out a bogus Stripe Gateway without a source fails at :payment
-      # Shipments and payments should then be cleared before rendering checkout
-      expect(response.status).to be 400
-      expect(flash[:error])
-        .to eq 'Payment could not be processed, please check the details you entered'
-      order.reload
-      expect(order.shipments.count).to be 0
-      expect(order.payments.count).to be 0
-      expect(order.adjustment_total).to eq 0
```

Comment:
Do we need to cover this scenario for the new checkout ?

Reply:
> Do we need to cover this scenario for the new checkout ?

Thanks, indeed this is the question I wanted to raise. I don't know if we reach this state on the current checkout - a (Stripe) payment entity without a source. If we do, then I think we should expect a status 400, and not 302. If we don't, then I guess there is no point having this new test.

For https://github.com/openfoodfoundation/openfoodnetwork/issues/10958 I'm trying to find a correspondence between old and new specs. I might run into such situations more often.

Reply:
Ok, I've added a commit to undo changes to this file, as the test is perhaps nit particularly useful.

Reply:
> Thanks, indeed this is the question I wanted to raise. I don't know if we reach this state on the current checkout - a (Stripe) payment entity without a source. If we do, then I think we should expect a status 400, and not 302. If we don't, then I guess there is no point having this new test.

I don't know enough about payment and source to know if it's plausible scenario, happy to have a look if needed. I am wondering if the new checkout does not return a 400 in this scenario but display an error to the user instead.

Reply:
It sounds like we should try to replicate it with a feature spec. If not replicable, then I guess we can move on.

Reply:
It's tough to replicate Stripe errors on feature specs - we'd need a proxy to intercept HTTP requests to the real Stripe... I've tried this with puffing billy but did not succeed.

This [error just came in on bugsnag](https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/66329690f4b6380007e8a4f8?event_id=6632969000e6f40fb4000000), it looks somewhat related (touches the checkout controller and the payment source). So maybe this scenario is relevant after all. I'll have a more thorough look tomorrow.

Reply:
Good pickup on the error!

I noticed that some data was blank in the request. Perhaps (hopefully) the credit card details were scrubbed before logging, but I wonder about `gateway_payment_profile_id`.

```
{
  "_method": "put",
...
  "order": {
    "payments_attributes": [
      {
        "source_attributes": {
...
          "month": "",
          "year": "",
          "cc_type": "",
          "last_digits": "",
          "gateway_payment_profile_id": ""
        },
        "payment_method_id": "201053"
      }
    ]
  },
  "commit": "Next - Order summary"
}
```

It gets set [a few different ways](https://github.com/search?q=repo%3Aopenfoodfoundation%2Fopenfoodnetwork+gateway_payment_profile_id+NOT+path%3A%2F%5Espec%5C%2F%2F+NOT+path%3Adb&type=code) in the codebase, but perhaps it's getting lost somehow.

Reply:
Interesting. The request points to `"payment_method_id": "201053"` which - by looking at the production DB using Metabase - leads to a `Spree::PaymentMethod::Check` payment method. It seems we're getting this error across production servers (an error 500 on checkout :eyes: ). Below, also on FR-prod:

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/49817236/acb66e01-f3ba-4abd-8d7d-c3bbdf118d9f)

One our codebase, we have [the line](https://github.com/openfoodfoundation/openfoodnetwork/blob/2e004a026b8be846d12075382b201ab56b0df8c7/app/models/spree/payment_method.rb#L83):
`    # nil means the payment method doesn't require a source e.g. check`

I wonder if this happens due to a sort of race condition in which initially a Stripe payment method is selected for checkout but, during this time, disabled but the admin... or so. Since it relates to checkout, I think it might deserve an issue.

Reply:
Ok, I've tried to reproduce the bugsnag error with a spec:
- with a cash payment method, we can trigger the error message, but somehow we get a status 200 (I was expecting 500')
- with Stripe payment method, we can see the data is incomplete, and just get the expected 422, with no error message

I'm not sure how to proceed on this one; do we expect payment_source_class to be required for payment methods with `type: "Spree::PaymentMethod::Check"`?

Reply:
Good find Filipe! I think the comment is right, we should just make it `nil` in this case. I've updated the code and the spec to show the error is now avoided. How does that look?